File: ios/Sources/RynBridgeContacts/ContactsModule.swift
import Foundation
import RynBridge

public struct ContactsModule: BridgeModule, Sendable {
    public let name = "contacts"
    public let version = "0.1.0"
    public let actions: [String: ActionHandler]

    public init(provider: ContactsProvider) {
        actions = [
            "getContacts": { payload in
                let query = payload["query"]?.stringValue
                let limit = payload["limit"]?.intValue ?? 100
                let offset = payload["offset"]?.intValue ?? 0
                let contacts = try await provider.getContacts(query: query, limit: limit, offset: offset)
                return ["contacts": .array(contacts.map { .dictionary($0.toPayload()) })]
            },
            "getContact": { payload in
                guard let id = payload["id"]?.stringValue else {
                    throw RynBridgeError(code: .invalidMessage, message: "Missing required field: id")
                }
                let contact = try await provider.getContact(id: id)
                return contact.toPayload()
            },
            "createContact": { payload in
                guard let givenName = payload["givenName"]?.stringValue else {
                    throw RynBridgeError(code: .invalidMessage, message: "Missing required field: givenName")
                }
                guard let familyName = payload["familyName"]?.stringValue else {
                    throw RynBridgeError(code: .invalidMessage, message: "Missing required field: familyName")
                }
                let phoneNumbers = payload["phoneNumbers"]?.arrayValue?.compactMap { item -> (label: String, number: String)? in
                    guard let dict = item.dictionaryValue,
                          let label = dict["label"]?.stringValue,
                          let number = dict["number"]?.stringValue else { return nil }
                    return (label: label, number: number)
                } ?? []
                let emailAddresses = payload["emailAddresses"]?.arrayValue?.compactMap { item -> (label: String, address: String)? in
                    guard let dict = item.dictionaryValue,
                          let label = dict["label"]?.stringValue,
                          let address = dict["address"]?.stringValue else { return nil }
                    return (label: label, address: address)
                } ?? []
                let id = try await provider.createContact(givenName: givenName, familyName: familyName, phoneNumbers: phoneNumbers, emailAddresses: emailAddresses)
                return ["id": .string(id)]
            },
            "updateContact": { payload in
                guard let id = payload["id"]?.stringValue else {
                    throw RynBridgeError(code: .invalidMessage, message: "Missing required field: id")
                }
                let givenName = payload["givenName"]?.stringValue
                let familyName = payload["familyName"]?.stringValue
                let phoneNumbers = payload["phoneNumbers"]?.arrayValue?.compactMap { item -> (label: String, number: String)? in
                    guard let dict = item.dictionaryValue,
                          let label = dict["label"]?.stringValue,
                          let number = dict["number"]?.stringValue else { return nil }
                    return (label: label, number: number)
                }
                let emailAddresses = payload["emailAddresses"]?.arrayValue?.compactMap { item -> (label: String, address: String)? in
                    guard let dict = item.dictionaryValue,
                          let label = dict["label"]?.stringValue,
                          let address = dict["address"]?.stringValue else { return nil }
                    return (label: label, address: address)
                }
                try await provider.updateContact(id: id, givenName: givenName, familyName: familyName, phoneNumbers: phoneNumbers, emailAddresses: emailAddresses)
                return [:]
            },
            "deleteContact": { payload in
                guard let id = payload["id"]?.stringValue else {
                    throw RynBridgeError(code: .invalidMessage, message: "Missing required field: id")
                }
                try await provider.deleteContact(id: id)
                return [:]
            },
            "pickContact": { _ in
                let contact = try await provider.pickContact()
                if let contact {
                    return ["contact": .dictionary(contact.toPayload())]
                } else {
                    return ["contact": .null]
                }
            },
            "requestPermission": { _ in
                let granted = try await provider.requestPermission()
                return ["granted": .bool(granted)]
            },
            "getPermissionStatus": { _ in
                let status = try await provider.getPermissionStatus()
                return ["status": .string(status)]
            },
        ]
    }
}

File: android/contacts/src/main/kotlin/io/rynbridge/contacts/ContactsModule.kt
package io.rynbridge.contacts

import android.content.Context
import io.rynbridge.core.*

class ContactsModule(provider: ContactsProvider) : BridgeModule {
    constructor(context: Context) : this(DefaultContactsProvider(context))

    override val name = "contacts"
    override val version = "0.1.0"
    override val actions: Map<String, ActionHandler> = mapOf(
        "getContacts" to { payload ->
            val query = payload["query"]?.stringValue
            val limit = payload["limit"]?.intValue?.toInt() ?: 100
            val offset = payload["offset"]?.intValue?.toInt() ?: 0
            val contacts = provider.getContacts(query, limit, offset)
            mapOf("contacts" to BridgeValue.array(contacts.map { BridgeValue.dict(it.toPayload()) }))
        },
        "getContact" to { payload ->
            val id = payload["id"]?.stringValue
                ?: throw RynBridgeError(code = ErrorCode.INVALID_MESSAGE, message = "Missing required field: id")
            val contact = provider.getContact(id)
            contact.toPayload()
        },
        "createContact" to { payload ->
            val givenName = payload["givenName"]?.stringValue
                ?: throw RynBridgeError(code = ErrorCode.INVALID_MESSAGE, message = "Missing required field: givenName")
            val familyName = payload["familyName"]?.stringValue
                ?: throw RynBridgeError(code = ErrorCode.INVALID_MESSAGE, message = "Missing required field: familyName")
            val phoneNumbers = payload["phoneNumbers"]?.arrayValue?.mapNotNull { item ->
                val dict = item.dictionaryValue ?: return@mapNotNull null
                val label = dict["label"]?.stringValue ?: return@mapNotNull null
                val number = dict["number"]?.stringValue ?: return@mapNotNull null
                Pair(label, number)
            } ?: emptyList()
            val emailAddresses = payload["emailAddresses"]?.arrayValue?.mapNotNull { item ->
                val dict = item.dictionaryValue ?: return@mapNotNull null
                val label = dict["label"]?.stringValue ?: return@mapNotNull null
                val address = dict["address"]?.stringValue ?: return@mapNotNull null
                Pair(label, address)
            } ?: emptyList()
            val id = provider.createContact(givenName, familyName, phoneNumbers, emailAddresses)
            mapOf("id" to BridgeValue.string(id))
        },
        "updateContact" to { payload ->
            val id = payload["id"]?.stringValue
                ?: throw RynBridgeError(code = ErrorCode.INVALID_MESSAGE, message = "Missing required field: id")
            val givenName = payload["givenName"]?.stringValue
            val familyName = payload["familyName"]?.stringValue
            val phoneNumbers = payload["phoneNumbers"]?.arrayValue?.mapNotNull { item ->
                val dict = item.dictionaryValue ?: return@mapNotNull null
                val label = dict["label"]?.stringValue ?: return@mapNotNull null
                val number = dict["number"]?.stringValue ?: return@mapNotNull null
                Pair(label, number)
            }
            val emailAddresses = payload["emailAddresses"]?.arrayValue?.mapNotNull { item ->
                val dict = item.dictionaryValue ?: return@mapNotNull null
                val label = dict["label"]?.stringValue ?: return@mapNotNull null
                val address = dict["address"]?.stringValue ?: return@mapNotNull null
                Pair(label, address)
            }
            provider.updateContact(id, givenName, familyName, phoneNumbers, emailAddresses)
            emptyMap()
        },
        "deleteContact" to { payload ->
            val id = payload["id"]?.stringValue
                ?: throw RynBridgeError(code = ErrorCode.INVALID_MESSAGE, message = "Missing required field: id")
            provider.deleteContact(id)
            emptyMap()
        },
        "pickContact" to { _ ->
            val contact = provider.pickContact()
            if (contact != null) {
                mapOf("contact" to BridgeValue.dict(contact.toPayload()))
            } else {
                mapOf("contact" to BridgeValue.nullValue())
            }
        },
        "requestPermission" to { _ ->
            val granted = provider.requestPermission()
            mapOf("granted" to BridgeValue.bool(granted))
        },
        "getPermissionStatus" to { _ ->
            val status = provider.getPermissionStatus()
            mapOf("status" to BridgeValue.string(status))
        }
    )
}
